lib/translator/status.cc

```diff
@@ -0,0 +1,252 @@
+      result.ascq =
+          scsi_defs::AdditionalSenseCodeQualifier::kInternalTargetFailure;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedByRequest:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedSqDeletion:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedFailedFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kAbortedMissingFused:
+      result.status = scsi_defs::Status::kTaskAborted;
+      result.sense_key = scsi_defs::SenseKey::kAbortedCommand;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kInvalidNamespaceOrFormat:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc =
+          scsi_defs::AdditionalSenseCode::kAccessDeniedInvalidLuIdentifier;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kAccessDeniedInvalidLuIdentifier;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kLbaOutOfRange:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      result.asc = scsi_defs::AdditionalSenseCode::kLbaOutOfRange;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::kLbaOutOfRange;
+      break;
+    case nvme_defs::GenericCommandStatusCode::kNamespaceNotReady:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kNotReady;
+      result.asc = scsi_defs::AdditionalSenseCode::
+          kLogicalUnitNotReadyCauseNotReportable;
+      result.ascq = scsi_defs::AdditionalSenseCodeQualifier::
+          kLogicalUnitNotReadyCauseNotReportable;
+      break;
+    default:
+      DebugLog("No SCSI translation for NVMe status with code %#x",
+               static_cast<uint8_t>(sc));
+      return StatusCode::kNoTranslation;
+  }
+  return StatusCode::kSuccess;
+}
+
+// Section 7.2
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode Status(nvme_defs::CommandSpecificStatusCode sc, ScsiStatus& result) {
+  result.asc = scsi_defs::AdditionalSenseCode::kNoAdditionalSenseInfo;
+  result.ascq = scsi_defs::AdditionalSenseCodeQualifier::kNoAdditionalSenseInfo;
+
+  switch (sc) {
+    case nvme_defs::CommandSpecificStatusCode::kCompletionQueueInvalid:
+      result.status = scsi_defs::Status::kCheckCondition;
+      result.sense_key = scsi_defs::SenseKey::kIllegalRequest;
+      break;
+    case nvme_defs::CommandSpecificStatusCode::
```

Comment:
These line breaks look awkward to me; would it make sense to put the entire identifier on one line and the "unspecified" comments above/below?

Reply:
I agree the one line approach looks cleaner. However, the git-clang-format automatically reformats it to be on two lines. We could potentially increase the maximum line length for clang-format (if possible) to get around this

Reply:
Ah, I meant possibly moving the // unspecified comments to not be every line; if you've got a big fall-through block, one comment is probably enough.

Reply:
Ah gotcha! Yep, I think doing so would still ensure it's clear that those statuses are unspecified and will look cleaner